### ISSUE TITLE
Model gives inaccurate results post conversion to tflite

### ISSUE THREAD

**Issue by AD-lite24** [PINTO0309/onnx2tf#685]
### Issue Type

Others

### OS

Linux

### onnx2tf version number

1.25.7

### onnx version number

1.16.2

### onnxruntime version number

1.18.1

### onnxsim (onnx_simplifier) version number

0.4.36

### tensorflow version number

2.17.0

### Download URL for ONNX

https://huggingface.co/onnx-community/metric3d-vit-small/blob/main/onnx/model.onnx

### Parameter Replacement JSON

```json
N/A
```

### Description

1. To deploy a monodepth model on edge devices. R&D work and problem exploration. Massive impact since nothing except tflite seems to work with snapdragon SOCs

2. The model outputs are not correct at all. I did a lot of inspection and here are my findings

Here are the input details for the tflite conversion

```
input_details = [{
    'name': 'pixel_values',
    'index': 0,
    'shape': np.array([1, 480, 640, 3], dtype=np.int32),
    'shape_signature': np.array([1, 480, 640, 3], dtype=np.int32),
    'dtype': <class 'numpy.float32'>,
    'quantization': (0.0, 0),
    'quantization_parameters': {
        'scales': np.array([], dtype=np.float32),
        'zero_points': np.array([], dtype=np.int32),
        'quantized_dimension': 0
    },
    'sparsity_parameters': {}
}]

```

and here are the output details 

```

output_details = [
    {
        'name': 'Identity',
        'index': 1765,
        'shape': np.array([1, 476, 628], dtype=np.int32),
        'shape_signature': np.array([1, 476, 628], dtype=np.int32),
        'dtype': <class 'numpy.float32'>
        'quantization': (0.0, 0),
        'quantization_parameters': {
            'scales': np.array([], dtype=np.float32),
            'zero_points': np.array([], dtype=np.int32),
            'quantized_dimension': 0
        },
        'sparsity_parameters': {}
    },
    {
        'name': 'Identity_1',
        'index': 1785,
        'shape': np.array([1, 3, 476, 628], dtype=np.int32),
        'shape_signature': np.array([1, 3, 476, 628], dtype=np.int32),
        'dtype': <class 'numpy.float32'>,
        'quantization': (0.0, 0),
        'quantization_parameters': {
            'scales': np.array([], dtype=np.float32),
            'zero_points': np.array([], dtype=np.int32),
            'quantized_dimension': 0
        },
        'sparsity_parameters': {}
    },
    {
        'name': 'Identity_2',
        'index': 1784,
        'shape': np.array([1, 476, 628], dtype=np.int32),
        'shape_signature': np.array([1, 476, 628], dtype=np.int32),
        'dtype': <class 'numpy.float32'>,
        'quantization': (0.0, 0),
        'quantization_parameters': {
            'scales': np.array([], dtype=np.float32),
            'zero_points': np.array([], dtype=np.int32),
            'quantized_dimension': 0
        },
        'sparsity_parameters': {}
    }
]
```

Upon inspection of the onnx file, the onnx version has 3 outputs

* Predicted Depth: tensor: float32[batch_size,4*floor(3.5*floor(height/14)),4*floor(3.5*floor(width/14))]
* predicted_normal: tensor: float32[batch_size,3,4*floor(3.5*floor(height/14)),4*floor(3.5*floor(width/14))]
* normal_confidence: tensor: float32[batch_size,4*floor(3.5*floor(height/14)),4*floor(3.5*floor(width/14))]

So in the tflite file, Identity, Identity_1, and identity_2 corresponds to either one of these. For predicted depth, it could be either identity or identity_2 and I tried both of them but neither give accurate results at all.

Identity gives values in the range of [-1000, -5000] which does not seem accurate for either confidence or depth values while 
Identity_2 gives values in the range of [10, 50] which seems more reasonable but still not accurate.

I am not sure if I was supposed to follow any pre or post processing steps different from the onnx format. Tflite often has different steps but I don't exactly know what they are. 

[this](https://github.com/YvanYin/Metric3D/blob/main/onnx/metric3d_onnx_export.py) is an example for drawing inference from the onnx file which works absolutely fine. Is it the conversion process that broke it or is there something additional I need to do to fix the results?

**Comment by PINTO0309** [comment]
As I commented in the previous issue, if you find it troublesome to correct the Transpose, change the input resolution of the model to a fixed resolution. You should start by checking for precision errors yourself first by making sure to include the `-cotof` option when converting. onnx2tf is imperfect when it comes to converting dynamic tensors.

**Comment by AD-lite24** [comment]
Really sorry for this. Referring back to the previous issue, the model input resolution was fixed to [1, 3, 480, 640] so I believed the dynamic input size was no longer a problem. with -cotof I can see a pretty bad divergence with an abs error of 4340.93 in the final output. But from (this)[https://github.com/PINTO0309/onnx2tf/issues/681#issuecomment-2312391931] step I thought the input tensors are no longer dynamic and I am reshaping my images according to the fixed tensor input.

Regardless I do not wish to take up anymore of your time, I will try to figure out why the the values are diverging

**Comment by PINTO0309** [comment]
Using the `-cotof` option will probably tell you which ops are missing the conversion.

- e.g.
  ![image](https://github.com/user-attachments/assets/f6a5ad6f-bf32-432f-9546-887b3d2a3761)

**Comment by PINTO0309** [comment]
- Tests - OK
  ![image](https://github.com/user-attachments/assets/ab17ee9f-f331-4e31-a0a7-8a0362b124ce)

**Comment by PINTO0309** [comment]
Fix: https://github.com/PINTO0309/onnx2tf/releases/tag/1.25.9

**Comment by AD-lite24** [comment]
Ah that is amazing. Brilliant as always. Results have improved significantly compared to the gibberish output as before. The values are still not accurate though from real world tests and the high frequency features were missed by the converted file. 

Do you know how we can find out how the pre and post processing steps change upon conversion? Maybe the image needs normalisation as well or the output needs some sort of scaling? The onnx range of values were [4.7, 24.7] and the tflite range is [0.68631876, 11.6238165] and with metric depth such disparity matters. This does seem at odds with the calculated error of 1e-4 but I am not sure.

Anyway this is certainly nothing you should worry about, you have helped me a lot. Maybe the model simply cannot be converted with very high precision. Again thanks a lot for taking out your time and fixing these issues!

**Comment by PINTO0309** [comment]
I can't say anything for sure, just guessing how you plan to use the model, but here are some common patterns of loss of accuracy that can occur after conversion to TensorFlow:
1. If you eventually quantize to INT8/Float16 and use the model, accuracy is likely to degrade if the normalization process is performed at the beginning of the model. This is because the results vary greatly depending on the type of calibration performed during quantization, so including normalization in preprocessing does not necessarily result in a deterioration in accuracy.
![image](https://github.com/user-attachments/assets/7e036b4a-6d56-4d7d-ac8c-bc9b99e03fff)
2. Certain operations (OPs) have a significant divergence between TensorFlow's internal implementation and ONNX and PyTorch's internal implementation. This may involve minor effects such as the internal rounding of numbers being different between ONNX and TensorFlow, or it may be fatal, with a bug on the TensorFlow side having been left unfixed for a long time.
    - https://github.com/PINTO0309/onnx2tf?tab=readme-ov-file#12-if-the-accuracy-of-the-float32-model-degrades-significantly
    - https://github.com/PINTO0309/onnx2tf?tab=readme-ov-file#7-if-the-accuracy-of-the-int8-quantized-model-degrades-significantly
3. Even if you convert the formula into a completely compatible pattern, there may be a large error that cannot be tolerated. This is also an issue with the internal implementation of ONNX.
    - https://github.com/PINTO0309/onnx2tf?tab=readme-ov-file#13-problem-of-extremely-large-calculation-error-in-instancenormalization
4. I always check that the error of the final output converges to around `1e-4`, but you may want to check the errors for each OP that are checked by the `-cotof` option yourself. The reason why it is better to check the accuracy check results of `-cotof` again is that there may be problems with large errors being rounded or flattened in the OP processing of `Softmax` or `Convoloution`. However, since the structure of the ViT model is very large, the amount of accuracy check results output by the `-cotf` option becomes enormous, and it is very difficult to visually check the error check results of all OPs.

**Comment by AD-lite24** [comment]
> If you eventually quantize to INT8/Float16 and use the model, accuracy is likely to degrade if the normalization process is performed at the beginning of the model. This is because the results vary greatly depending on the type of calibration performed during quantization, so including normalization in preprocessing does not necessarily result in a deterioration in accuracy.

Makes sense. Using float32 here but yes still holds true.
 
Thanks for the detail, I get why conversion is so hard especially for these larger models. So from what I can tell it is just not that easy to convert accurately for this particular model. And any arbitrary changes made by tflite cant be predicted (though normalization is not a big factor). 

> Therefore, as you point out, the situation where the final output values ​​of each model differ by more than a factor of 10 is clearly not a problem with the models themselves.

So the issue is tflite? The values are close but not accurate enough, but that cant be explained since -cotof test does give an error of less than 1e-4. Maybe some depth scaling? I will figure out the scale factor if there is one. Thanks a bunch!

**Comment by PINTO0309** [comment]
I have to attend a conference for the next three days, so my investigation and definitive answer will be a little delayed.
